src/main/java/com/uber/cadence/internal/worker/Poller.java

```diff
@@ -267,14 +291,22 @@ private class PollExecutionTask implements Poller.ThrowingRunnable {
     @Override
     public void run() throws Exception {
       try {
-        pollSemaphore.acquire();
-        T task = pollTask.poll();
-        if (task == null) {
-          return;
+        pollerAutoScaler.acquire();
+        try {
+          pollSemaphore.acquire();
```

Comment:
Seems like original `pollSemaphore` does not participate in any meaningful way. It is created for every `PollExecutionTask` separately and is not shared. I suggest removing this entirely.

Reply:
Agree, removed it.